docs/interfaces.md

```diff
@@ -0,0 +1,431 @@
+
+// set with 
+namespaces/registered/<enclave_vk> -> CCKeyRegistrationMessage
+
+// set with exportCCKeys and retrieved using importCCKeys
+namespaces/exported/<namepsace>/<enclave_vk> -> ExportMessage
+```
+
+Data types are defined as follows:
+
+```go
+type Credentials struct {
+        Params CC_Parameters,   // cc_params as provided during enclave creation
+        Enclave_VK []byte,      // chaincode enclave public key
+        Evidence []byte,        // attestation evidence (epid/dcap/etc...)
+}
+
+type CCKeyRegistrationMessage struct {
+        Chaincode_EK []byte,    // public chaincode encryption key
+        Enclave_VK []byte,      // creator of this message
+        Sig_Enclave []byte,     // signature of the message creator
+}
+
+type ExportMessage struct {
+        Chaincode_EK []byte,    // public chaincode encryption key
+        CCKeys_enc []byte,      // chaincode keys encrypted for the receiver
+        Enclave2_VK []byte,     // creator of this message
+        Enclave_VK []byte,      // receiver of this message
+        Sig_Enclave []byte,     // signature of the message creator
+}
+```
+
+ERCC keeps in instance of an attestation.Verifier to verify the attestation evidence. In case of EPID, the evidence is an `IASAttestationReport` object. This instance has the Intel attestation verification key.
+```go
+type EnclaveRegistryCC struct {
+	ra  attestation.Verifier
+}
+
+// Intel IAS verification key (same as used by TLCC Enclave)
+const IntelPubPEM = `...`
+
+type IASAttestationReport struct {
+	EnclavePk                   []byte `json:"EnclavePk"`
+	IASReportSignature          string `json:"IASReport-Signature"`
+	IASReportSigningCertificate string `json:"IASReport-Signing-Certificate"`
+	IASReportBody               []byte `json:"IASResponseBody"`
+}
+```
+
+
+# TLCC
+
+## Interface:
+The TLCC interface is implemented through a "normal" chaincode interface.
+
+```go
+// chaincode interface 
+func Invoke(stub shim.ChaincodeStubInterface) pb.Response {}
+
+// TODO note that we are multiplexing normal and range queries with this method
+// used by ecc via cc2cc
+getMetadata(key string, nonce string, isRangeQuery bool) (cmacBase64  string, error)
+
+validateMSP(mspId string, nonce string) (valid bool, cmacBase64 string, error)
+
+// used by the admin to join the channel
+joinChannel(channel_id string) error
+```
+
+```c++
+// interface exposed to TLCC_enclave via cgo
+// logging
+void ocall_print_string([in, string] const char *str);
+```
+
+## State:
+TLCC keeps a local reference of the enclave id. This reference is volatile. Also a reference to the peer is kept to access the ledger and read blocks.
+
+```go
+type TrustedLedgerCC struct {
+	enclave enclave.Stub
+	peer    *peer.Peer
+}
+
+type StubImpl struct {
+	eid C.enclave_id_t
+}
+```
+
+
+# TLCC_Enclave
+
+## Interface:
+Defined in the TLCC EDL.
+```c++
+// TBD is this interface already sufficient to support multi-channel?
```

Comment:
hmm, don't think so. To some extent it depends on how you define key (which could be defined as the concatenation of `channel-id || chaincode-id || chaincode-key`) but i think having the triplet explicit is nicer/easier to deal with and also easier to understand.

Reply:
added namespace. as a single tcc_enclave is responsible for a single channel, the dispatching using the channel id is done already outside tlcc_enclave, thus, no need to add it tlcc.edl

Reply:
did we decide at some point that multi-channel means multi-enclave?  I guess it is now somewhat implicit in TLCC state. Maybe then worth also explicitly adding the channel id to the tlcc_enclave state?

Reply:
added the channel id and a hash of the genesis block